indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+      pendingRequests.addAll(allRequests);
+    } else {
+      final Interval[] sortedUsedSegmentIntervals = getSortedIntervals(usedSegments);
+      final Map<Interval, List<SegmentAllocateRequest>> usedIntervalToRequests = new HashMap<>();
+
+      for (SegmentAllocateRequest request : allRequests) {
+        // If there is an overlapping used segment interval, that interval is
+        // the only candidate for allocation
+        Interval overlappingInterval = findOverlappingInterval(
+            request.getRowInterval(),
+            sortedUsedSegmentIntervals
+        );
+
+        if (overlappingInterval == null) {
+          pendingRequests.add(request);
+        } else if (overlappingInterval.contains(request.getRowInterval())) {
+          // Found an enclosing interval, use this for allocation
+          usedIntervalToRequests.computeIfAbsent(overlappingInterval, i -> new ArrayList<>())
+                                .add(request);
+        }
+      }
+
+      // Try to allocate segments for the identified used segment intervals
+      // Do not retry the failed requests with other intervals unless the batch is requeued
+      for (Map.Entry<Interval, List<SegmentAllocateRequest>> entry : usedIntervalToRequests.entrySet()) {
+        List<SegmentAllocateRequest> successfulRequests = allocateSegmentsForInterval(
+            entry.getKey(),
+            entry.getValue(),
+            requestBatch
+        );
+        successCount += successfulRequests.size();
+      }
+    }
+
+    // For requests that do not overlap with a used segment, first try to allocate
+    // using the preferred granularity, then smaller granularities
+    for (Granularity granularity :
+        Granularity.granularitiesFinerThan(requestBatch.key.preferredSegmentGranularity)) {
+      Map<Interval, List<SegmentAllocateRequest>> requestsByInterval =
+          getRequestsByInterval(pendingRequests, granularity);
+
+      for (Map.Entry<Interval, List<SegmentAllocateRequest>> entry : requestsByInterval.entrySet()) {
+        List<SegmentAllocateRequest> successfulRequests = allocateSegmentsForInterval(
+            entry.getKey(),
+            entry.getValue(),
+            requestBatch
+        );
+        successCount += successfulRequests.size();
+        pendingRequests.removeAll(successfulRequests);
+      }
+    }
+
+    return successCount;
+  }
+
+  private Interval findOverlappingInterval(Interval searchInterval, Interval[] sortedIntervals)
+  {
+    int index = Arrays.binarySearch(
+        sortedIntervals,
+        searchInterval,
+        Comparators.intervalsByStartThenEnd()
+    );
+    if (index >= 0) {
+      return sortedIntervals[index];
+    }
+
+    // Key was not found, returned index is (-(insertionPoint) - 1)
+    index = -(index + 1);
+
+    // If the interval at index doesn't overlap, (index + 1) wouldn't overlap either
+    if (index < sortedIntervals.length) {
+      if (sortedIntervals[index].overlaps(searchInterval)) {
```

Comment:
index can be negative too.

Reply:
Handled in line 366. I think a test for this method would be helpful.